engine/apps/grafana_plugin/views/self_hosted_install.py

```diff
@@ -1,43 +1,48 @@
+from apps.grafana_plugin.helpers import GrafanaAPIClient
+from apps.user_management.models.organization import Organization, ProvisionedPlugin
 from apps.user_management.sync import sync_organization
 from common.api_helpers.mixins import GrafanaHeadersMixin
 
 
 class SelfHostedInstallView(GrafanaHeadersMixin, APIView):
-    permission_classes = (SelfHostedInvitationTokenVerified,)
-
-    def remove_invitation_token(self, token):
-        DynamicSetting = apps.get_model("base", "DynamicSetting")
-        self_hosted_settings = DynamicSetting.objects.get_or_create(
-            name="self_hosted_invitations",
-            defaults={
-                "json_value": {
-                    "keys": [],
-                }
-            },
-        )[0]
-        self_hosted_settings.json_value["keys"].remove(token)
-        self_hosted_settings.save(update_fields=["json_value"])
-
-    def post(self, request: Request) -> Response:
-        token_string = get_authorization_header(request).decode()
+    def post(self, _request: Request) -> Response:
+        """
+        We've already validated that settings.GRAFANA_API_URL is set (in apps.grafana_plugin.GrafanaPluginConfig)
+        The user is now trying to finish plugin installation. We'll take the Grafana API url that they specified +
+        the token that we are provided and first verify them. If all is good, upsert the organization in the database,
+        and provision the plugin.
+        """
         stack_id = settings.SELF_HOSTED_SETTINGS["STACK_ID"]
         org_id = settings.SELF_HOSTED_SETTINGS["ORG_ID"]
+        grafana_url = settings.SELF_HOSTED_SETTINGS["GRAFANA_API_URL"]
+        grafana_api_token = self.instance_context["grafana_token"]
+
+        grafana_api_client = GrafanaAPIClient(api_url=grafana_url, api_token=grafana_api_token)
+        _, client_status = grafana_api_client.check_token()
+        status_code = client_status["status_code"]
+
+        provisioning_info: ProvisionedPlugin = {"error": None}
+
+        if status_code == status.HTTP_404_NOT_FOUND:
+            provisioning_info["error"] = f"Unable to connect to the specified Grafana API - {grafana_url}"
+            return Response(data=provisioning_info, status=status.HTTP_400_BAD_REQUEST)
+        elif status_code in [status.HTTP_401_UNAUTHORIZED, status.HTTP_403_FORBIDDEN]:
+            provisioning_info[
+                "error"
+            ] = f"You are not authorized to communicate with the specified Grafana API - {grafana_url}"
+            return Response(data=provisioning_info, status=status.HTTP_400_BAD_REQUEST)
 
         organization = Organization.objects.filter(stack_id=stack_id, org_id=org_id).first()
         if organization:
             organization.revoke_plugin()
-            organization.grafana_url = self.instance_context["grafana_url"]
-            organization.api_token = self.instance_context["grafana_token"]
+            organization.grafana_url = grafana_url
```

Comment:
if we take the grafana url from settings.SELF_HOSTED_SETTINGS, which http://grafana:3000 by default (https://github.com/grafana/oncall/pull/624/files#diff-f5d10b03472abe3719098ae8a8855468e92524ebe790c39a34d2c632f3f0486dR23), the rendered urls to grafana will be broken. If self.instance_context["grafana_url"] returns public url we should use it, otherwise we should specify public url in settings.SELF_HOSTED_SETTINGS

Reply:
I think I understand what you mean. The only way to get this to work then would to be something like this:
```diff
x-env-vars: &oncall-env-vars
  BROKER_TYPE: ${BROKER_TYPE}
-  GRAFANA_API_URL: http://grafana:3000
+  GRAFANA_API_URL: http://localhost:3000
```
this would generate public URLs that point to the actual local grafana, but the problem is that the OnCall containers cannot communicate w/ the `grafana` container @ `localhost:3000`. Maybe there is some docker networking magic that I can add to the oncall containers to make `localhost:3000` point to `grafana:3000`?

Reply:
so one solution that works that I just tested is to modify `/etc/hosts` to add the following:
```
127.0.0.1 grafana
```
I could modify `make start` to idempotently add this record to that file, thoughts?

Reply:
@iskhakov see [this commit](https://github.com/grafana/oncall/pull/624/commits/e39d116444f22bed12abb8dc35e9308de1be5a92), lmk your thoughts. I know it's not ideal having to modify `/etc/hosts`, but the [`hostess` tool](https://github.com/cbednarski/hostess) I introduced in the `Makefile` makes it quite simple and cross-platform compatible.

Alternative solution, like we discussed, is to use `http://localhost:3000` as the value for `GRAFANA_API_URL` and _somehow_ modify the `oncall_engine`/`oncall_celery` containers such that within those containers `http://localhost:3000` points to `http://grafana:3000`. I have yet to find a solution for this.